### PR TITLE
Implement pretty print for CustomCallOp, ReducePrecisionOp

### LINKED ISSUE #20: Enhance ReducePrecisionOp, CustomCallOp printing
A separate issue will be used to track operations with single attributes, as those can be enhanced in an unambiguous manner relatively easily. These operations are a few of the ones with multiple attributes that can be easily enhanced.

Current proposal is to model `custom_call` after `func.call`, except using quotes since the call target can be an arbitrary string. 

```
%1 = "mhlo.custom_call"(%arg0, %arg1) {backend_config = "bar", call_target_name = "foo", has_side_effect = true} : (tensor<2x3xf32>, tensor<5x5xf32>) -> tensor<1x2x3xf32>
==>
%1 = mhlo.custom_call "foo"(%arg0, %arg1) {backend_config = "bar", has_side_effect = true} : (tensor<2x3xf32>, tensor<5x5xf32>) -> tensor<1x2x3xf32>
```

Use `e#m#` shorthand for exponent mantissa in `reduce_precision`, as seems to be common in chat messages on the topic:

```
%0 = "mhlo.reduce_precision"(%arg0) {exponent_bits = 8 : i32, mantissa_bits = 10 : i32} : (tensor<3x4xf32>) -> tensor<3x4xf32>
===>
%0 = mhlo.reduce_precision %arg0, e8m10 : tensor<3x4xf32>
```

### PR DESCRIPTION
**Pull request by GleasonK**
Print for CustomCallOp is based on [func.call](https://mlir.llvm.org/docs/Dialects/Func/#funccall-mlirfunccallop):

```
%1 = func.call @my_add(%0, %1) : (f32, f32) -> f32
```

The main difference is that quotes must remain around the call target since it may be an arbitrary string:

```
%1 = mhlo.custom_call "my_add"(%arg0, %arg1) {backend_config = "bar", has_side_effect = true} : (tensor<2x3xf32>, tensor<5x5xf32>) -> tensor<1x2x3xf32>
```

The print for ReducePrecisionOp is uses an `e#m#` format specifier, which is a common way to refer to FP precision of exponent/mantissa.

```
%0 = "stablehlo.reduce_precision"(%arg0) {exponent_bits = 8 : i32, mantissa_bits = 10 : i32} : (tensor<3x4xf32>) -> tensor<3x4xf32>
 ==>
%0 = stablehlo.reduce_precision %arg0, format = e8m10 : tensor<3x4xf32>
```

Added tests for invalid formats of `e#m#` as well as pretty prints of ops.

Closes #20

### CODE REVIEW

File: stablehlo/dialect/StablehloOps.td [hunk 1]
```diff
@@ -2627,6 +2640,12 @@ def StableHLO_ReducePrecisionOp :
     implementations.
 
     See https://www.tensorflow.org/xla/operation_semantics#reduceprecision.
+
+    Example:
+
+    ```mlir
+    %0 = mhlo.reduce_precision %arg0, format = e8m10 : tensor<3x4xf32>
```
Comment (burmako) [inline stablehlo/dialect/StablehloOps.td:2647]: Looks like PR description should be updated to use the new assemblyFormat?

File: stablehlo/tests/print_stablehlo.mlir [hunk 2]
```diff
@@ -208,4 +217,4 @@ func.func @encodings(%arg0: tensor<10x20xf32, #CSR>,
   %2 = "stablehlo.abs"(%arg0) : (tensor<10x20xf32, #CSR>) -> tensor<10x20xf32>
   %3 = "stablehlo.abs"(%arg0) : (tensor<10x20xf32, #CSR>) -> tensor<10x20xf32, #CSR>
   func.return %0 : tensor<10x20xf32>
-}
+}
```
Comment (burmako) [inline stablehlo/tests/print_stablehlo.mlir:220]: Typo: no newline at the end of file.

File: stablehlo/tests/print_types_invalid.mlir [hunk 3]
```diff
@@ -85,3 +85,59 @@ func.func @one_result_type(%arg0: tensor<1xf64>) -> tensor<1xf64> {
...
+func.func @reduce_precision_not_literal(%arg0: tensor<3x4xf32>) -> (tensor<3x4xf32>) {
+  // expected-error @+1 {{custom op 'stablehlo.reduce_precision' expected valid keyword}}
+  %0 = stablehlo.reduce_precision %arg0, format = "e2m2" : tensor<3x4xf32>
+  func.return %0 : tensor<?x?xf64>
+}
+
+// -----
+
+func.func @reduce_precision_no_em(%arg0: tensor<3x4xf32>) -> (tensor<3x4xf32>) {
+  // expected-error @+1 {{custom op 'stablehlo.reduce_precision' expected exponent mantissa in format e#m#, saw z4f2}}
```
Comment (burmako) [inline stablehlo/tests/print_types_invalid.mlir:100]: As I'm looking through StablehloOps.cpp, I see that we have a bunch of different ways to express expected vs actual mismatches: "expected, saw", "expected, got", "expected, found", etc. We'll need to sort this out at some later point.

File: stablehlo/dialect/StablehloOps.cpp [hunk 4]
```diff
@@ -3591,6 +3591,64 @@ Operation* ReduceWindowOp::getReductionOp(int resultIndex) {
...
+  }
+
+  // Parse as base 10 integer strings
+  int exp, mant;
+  if (expS.getAsInteger(/*radix=*/10, exp) ||
+      manS.getAsInteger(/*radix=*/10, mant)) {
+    parser.emitError(loc, "unable to parse exponent or mantissa")
+        << expS.str() << ", " << manS.str();
+    return failure();
+  }
```
Comment (burmako) [inline stablehlo/dialect/StablehloOps.cpp:3644]: How about using `llvm::Regex` instead of substring-based parsing?

**Review by burmako — CHANGES_REQUESTED** [review]
Looks like I clicked "Comment" in my previous review, instead of "Request changes".

File: stablehlo/tests/print_types_invalid.mlir (hunk 3 above)
Comment (GleasonK) [reply stablehlo/tests/print_types_invalid.mlir:100]: Agree. Created #63.

File: stablehlo/dialect/StablehloOps.cpp (hunk 4 above)
Comment (GleasonK) [reply stablehlo/dialect/StablehloOps.cpp:3640]: Modified to use regex.

File: stablehlo/dialect/StablehloOps.td (hunk 1 above)
Comment (GleasonK) [reply stablehlo/dialect/StablehloOps.td]: PR description has `%0 = stablehlo.reduce_precision %arg0, format = e8m10 : tensor<3x4xf32>`.

That said, this line says `mhlo` which is incorrect. Fixed all uses of mhlo in doc examples.

File: stablehlo/dialect/StablehloOps.cpp [hunk 5]
```diff
@@ -3591,6 +3591,59 @@ Operation* ReduceWindowOp::getReductionOp(int resultIndex) {
...
+  llvm::StringRef manS = matches[2];
+
+  // Parse as base 10 integer strings
+  int exp, mant;
+  if (expS.getAsInteger(/*radix=*/10, exp) ||
+      manS.getAsInteger(/*radix=*/10, mant)) {
+    parser.emitError(loc, "unable to parse exponent or mantissa ")
+        << expS.str() << ", " << manS.str();
+    return failure();
+  }
```
Comment (burmako) [inline stablehlo/dialect/StablehloOps.cpp:3639]: Nit: split into two error messages? You have two tests for that anyway :)

**Review by burmako — APPROVED** [review]

File: stablehlo/dialect/StablehloOps.cpp (hunk 5 above)
Comment (GleasonK) [reply stablehlo/dialect/StablehloOps.cpp:3639]: split into two

**Pull request merged**